File: problems/data_structures/heap/lc_1705.py
from typing import List
import heapq


# fmt: off
class Solution:
    def eatenApples(self, apples: List[int], days: List[int]) -> int:
        ans = 0
        minq = []
        n = len(apples)
        i = 0
        while i < n or minq:
            while minq and minq[0][0] <= i: # rot
                heapq.heappop(minq)
            if i < n and apples[i]:
                heapq.heappush(minq, [i + days[i], apples[i]])
            if minq:
                ans += 1
                minq[0][1] -= 1
                if minq[0][1] == 0:
                    heapq.heappop(minq)
            i += 1
        return ans
# fmt: on


apples = [1, 2, 3, 5, 2]
days = [3, 2, 1, 4, 2]
assert Solution().eatenApples(apples, days) == 7

apples = [3, 0, 0, 0, 0, 2]
days = [3, 0, 0, 0, 0, 2]
assert Solution().eatenApples(apples, days) == 5

File: problems/data_structures/heap/lc_1705.cpp
#include <functional>
#include <map>
#include <queue>
#include <utility>
#include <vector>
using std::greater;
using std::map;
using std::pair;
using std::priority_queue;
using std::vector;


// clang-format off

/**
 * std::priority_queue
 *     https://en.cppreference.com/w/cpp/container/priority_queue
 */
class Solution {
public:
  int eatenApples(vector<int> &apples, vector<int> &days) {
    int ans = 0;
    int n = apples.size();
    int i = 0;
    priority_queue<pair<int, int>, vector<pair<int, int>>, greater<>>
            minq;
    while (i < n || !minq.empty()) {
      while (!minq.empty() && minq.top().first <= i) minq.pop();
      if (i < n && apples[i] > 0) minq.emplace(i + days[i], apples[i]);
      if (!minq.empty()) {
        ++ans;
        auto [ddl, num] = minq.top(); minq.pop();
        --num;
        if (num > 0) minq.emplace(ddl, num);
      }
      ++i;
    }
    return ans;
  }
};
// clang-format on